dist/bin/scala

```diff
@@ -27,147 +26,211 @@ if [ -z "$PROG_HOME" ] ; then
   cd "$saveddir"
 fi
 
-addJvmOptions () {
-  jvm_options+=("'$1'")
-}
+source "$PROG_HOME/bin/common"
 
-addScalacOptions () {
-  java_options+=("'$1'")
+# This script operates in one of 3 usage modes:
+#   script
+#   repl
+#   run
+# execute_mode replaces mutually exclusive booleans:
+#  execute_repl=false
+#  execute_run=false
+#  execute_script=false
+setExecuteMode () {
+  case "${execute_mode-}" in
+  "") execute_mode="$1" ; shift ;;
+  *) echo "execute_mode==[${execute_mode-}], attempted overwrite by [$1]" 1>&2
+    exit 1
+    ;;
+  esac
 }
 
-source "$PROG_HOME/bin/common"
 
-declare -a residual_args
-declare -a script_args
-execute_repl=false
-execute_run=false
-execute_script=false
-with_compiler=false
-class_path_count=0
-CLASS_PATH=""
-save_compiled=false
+with_compiler=false             # to add compiler jars to repl classpath
+let class_path_count=0 || true  # count classpath args, warning if more than 1
```

Comment:
Just curious: what is the purpose of `|| true`?

Reply:
> Just curious: what is the purpose of `|| true`?

To manually verify no unintentional errors (typo's, etc), I temporarily add a couple of settings to the top of the file and manually exercise the script:
```bash
set -o errexit
set -o nounset
```
With `nounset`, it becomes necessary to deal with legitimately unset variables:
Any variable that is validly unset changed from `"$DEBUG"` to `"${DEBUG-}"`, and
unsetting of variables followed by `|| true`.

We could make the more strict temporary settings permanent, but there might be a risk of introducing a needless crash in some environment.